src/jail_plugins.c

```diff
@@ -0,0 +1,328 @@
+ *   3. Neither the names of the copyright holders nor the names of its
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation
+ * are those of the authors and should not be interpreted as representing
+ * any official policies, either expressed or implied.
+ */
+
+#include "jail_plugins.h"
+#include "jail_plugin_p.h"
+#include "jail_plugin_module.h"
+#include "jail_version.h"
+
+#include <gutil_log.h>
+#include <gutil_strv.h>
+
+#include <stdlib.h>
+#include <dlfcn.h>
+
+struct jail_plugins {
+    GSList* plugins;
+};
+
+static
+GStrV*
+jail_plugins_scan_plugin_dir(
+    const char* plugin_dir)
+{
+    GStrV* files = NULL;
+    GDir* dir = g_dir_open(plugin_dir, 0, NULL);
+
+    if (dir) {
+	const char* file;
+
+        while ((file = g_dir_read_name(dir)) != NULL) {
+            if (!g_str_has_prefix(file, "lib") &&
+                g_str_has_suffix(file, ".so")) {
+                files = gutil_strv_add(files, file);
+            }
+        }
+        g_dir_close(dir);
+
+        /* Sort file names to guarantee precedence order in case of conflict */
+        gutil_strv_sort(files, TRUE);
+    }
+    return files;
+}
+
+static
+JailPlugin*
+jail_plugins_find(
+    JailPlugins* self,
+    const char* name)
+{
+    GSList* l = self->plugins;
+
+    for (l = self->plugins; l; l = l->next) {
+        JailPlugin* plugin = l->data;
+        const JailPluginDesc* desc = plugin->desc;
+
+        if (!g_strcmp0(desc->name, name)) {
+            return plugin;
+        }
+    }
+    return NULL;
+}
+
+static
+gboolean
+jail_plugins_init_module(
+    JailPlugins* self,
+    const JailPluginModule* module,
+    const char* path,
+    const GStrV* enable,
+    const GStrV* disable)
+{
+    int n = 0;
+    const JailPluginDesc** p = module->plugins_start;
+
+    while (p <  module->plugins_end) {
+        const JailPluginDesc* desc = *p++;
+
+        if (desc->magic != JAIL_PLUGIN_MAGIC || !desc->name || !desc->create) {
+            GWARN("Invalid plugin \"%s\" in %s (ignored)", desc->name, path);
```

Comment:
desc->name can be null, does GWARN() tolerate that?

Reply:
It's pretty safe (should be turned into `"(null)"`) but I'll think about handling this case separately...